src/main/scala/firrtl/FirrtlExecutionOptions.scala

```diff
@@ -0,0 +1,752 @@
+    if (emitOneFilePerModule) { OneFilePerModule(optionsManager.targetDirName)                                    }
+    else                      { SingleFile(optionsManager.getBuildFileName(outputSuffix, outputFileNameOverride)) }
+
+  /** Get the user-specified targetFile assuming [[OutputConfig]] is [[SingleFile]]
+    *
+    * @param optionsManager this is needed to access build function and its common options
+    * @return the targetFile as a String
+    */
+  @deprecated("Use ExecutionOptionsManager.getTargetFile", "1.2.0")
+  def getTargetFile(optionsManager: ExecutionOptionsManager with HasFirrtlExecutionOptions): String =
+    getOutputConfig(optionsManager) match {
+      case SingleFile(targetFile) => targetFile
+      case _                      => throw new Exception("OutputConfig is not SingleFile!") }
+}
+
+/** A result of running the FIRRTL compiler */
+sealed trait FirrtlExecutionResult
+
+/** A successful result from running the FIRRTL compiler */
+object FirrtlExecutionSuccess {
+
+  /** Create a new successful execution
+    *
+    * @param emitType the compiler name used
+    * @param emitted the resulting FIRRTL circuit (as a [[scala.String]]
+    * @param circuitState the final [[CircuitState]]
+    * @return a [[FirrtlExecutionSuccess]] object
+    */
+  def apply(
+    emitType    : String,
+    emitted     : String,
+    circuitState: CircuitState
+  ): FirrtlExecutionSuccess = new FirrtlExecutionSuccess(emitType, emitted, circuitState)
+
+
+  /** Extractor to get at the compiler name and resulting FIRRTL */
+  def unapply(arg: FirrtlExecutionSuccess): Option[(String, String)] = {
+    Some((arg.emitType, arg.emitted))
+  }
+}
+/** A successful result from running the FIRRTL compiler
+  *
+  * @param emitType the name of the compiler used, currently "high", "middle", "low", "verilog", or "sverilog"
+  * @param emitted the emitted result of compilation
+  * @param circuitState the resulting [[CircuitState]]
+  */
+class FirrtlExecutionSuccess(
+  val emitType: String,
+  val emitted : String,
+  val circuitState: CircuitState
+) extends FirrtlExecutionResult
+
+/** An _unsuccessful_ result from running the FIRRTL compiler
+  *
+  * @param message some kind of hint as to what went wrong.
+  */
+case class FirrtlExecutionFailure(message: String) extends FirrtlExecutionResult
+
+/** Utilities that help with processing FIRRTL options */
+object FirrtlExecutionUtils {
+  /** Determine the target directory with the following precedence (highest to lowest):
+    *  - Explicitly from the user-specified `--target-dir`
+    *  - Implicitly from the default of [[FirrtlExecutionOptions.targetDirName]]
+    *
+    * @param annotations input annotations to extract targetDir from
+    * @return the target directory
+    * @note this is safe to use before [[HasFirrtlExecutionOptions.firrtlOptions]] is set
+    */
+  def targetDir(annotations: Seq[Annotation]): String = annotations
+    .reverse
```

Comment:
Why is this reversing?

Reply:
This requires some more thought... but, I was intending that the last `TargetDirAnnotation` should win. However, this conflicts with how the validation of the command line options is done. You can't specify `--target-dir` more than once. It would seem logical that this same check should be applied to the associated annotation.